### PR TITLE
Merge branch

### PR DESCRIPTION
**Pull request by haipengl**
merging branch with OjaNewton

### CODE REVIEW

File: cs/cs/cs.csproj [hunk 1]
```diff
@@ -1,5 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.props" Condition="Exists('..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.props')" />
```
Comment (JohnLangford) [inline cs/cs/cs.csproj]: Delete this?

File: cs/cs/cs.csproj [hunk 2]
```diff
@@ -16,6 +17,7 @@
     </TargetFrameworkProfile>
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\vowpalwabbit\</SolutionDir>
     <CodeContractsAssemblyMode>0</CodeContractsAssemblyMode>
+    <NuGetPackageImportStamp>884b164c</NuGetPackageImportStamp>
```
Comment (JohnLangford) [inline cs/cs/cs.csproj]: Delete?

File: cs/cs/cs.csproj [hunk 3]
```diff
@@ -100,8 +102,18 @@
     <CodeAnalysisRuleSet>MinimumRecommendedRules.ruleset</CodeAnalysisRuleSet>
   </PropertyGroup>
   <ItemGroup>
+<<<<<<< HEAD:cs/cs.csproj
```
Comment (JohnLangford) [inline cs/cs/cs.csproj]: Delete?

File: cs/cs/cs.csproj [hunk 4]
```diff
@@ -100,8 +102,18 @@
...
+      <HintPath>..\vowpalwabbit\packages\Antlr4.Runtime.4.3.0\lib\net45\Antlr4.Runtime.net45.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="System" />
+    <Reference Include="System.Core" />
+    <Reference Include="System.Threading.Tasks.Dataflow, Version=4.5.24.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+=======
     <Reference Include="GitLink, Version=2.2.0.0, Culture=neutral, processorArchitecture=MSIL">
       <HintPath>..\..\vowpalwabbit\packages\gitlink.2.2.0\lib\net45\GitLink.exe</HintPath>
+>>>>>>> 684798fdd2eceb9abb108fe9e64a0c4ef5ef4a07:cs/cs/cs.csproj
```
Comment (JohnLangford) [inline cs/cs/cs.csproj]: Delete?

File: cs/cs/cs.csproj [hunk 5]
```diff
@@ -189,10 +201,13 @@
       <ErrorText>This project references NuGet package(s) that are missing on this computer. Enable NuGet Package Restore to download them. For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
     </PropertyGroup>
     <Error Condition="!Exists('$(SolutionDir)\.nuget\NuGet.targets')" Text="$([System.String]::Format('$(ErrorText)', '$(SolutionDir)\.nuget\NuGet.targets'))" />
+    <Error Condition="!Exists('..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.props')" Text="$([System.String]::Format('$(ErrorText)', '..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.props'))" />
```
Comment (JohnLangford) [inline cs/cs/cs.csproj]: Delete?

File: cs/cs/cs.csproj [hunk 6]
```diff
@@ -189,10 +201,13 @@
...
+    <Error Condition="!Exists('..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.targets'))" />
   </Target>
   <PropertyGroup>
     <PreBuildEvent>$(SolutionDir)..\cs\transform_all.bat $(ProjectDir)</PreBuildEvent>
   </PropertyGroup>
+  <Import Project="..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.targets" Condition="Exists('..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.targets')" />
```
Comment (JohnLangford) [inline cs/cs/cs.csproj]: Delete?

File: cs/cs_parallel/packages.config [hunk 7]
```diff
@@ -1,5 +1,10 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+<<<<<<< HEAD:cs/packages.config
```
Comment (JohnLangford) [inline cs/cs_parallel/packages.config]: Delete?

File: cs/cs_parallel/packages.config [hunk 8]
```diff
@@ -1,5 +1,10 @@
...
+  <package id="Antlr4" version="4.3.0" targetFramework="net45" />
+  <package id="Antlr4.Runtime" version="4.3.0" targetFramework="net45" />
+=======
   <package id="gitlink" version="2.2.0" targetFramework="net45" />
+>>>>>>> 684798fdd2eceb9abb108fe9e64a0c4ef5ef4a07:cs/cs_parallel/packages.config
```
Comment (JohnLangford) [inline cs/cs_parallel/packages.config]: Delete?

File: cs/test/cs_test.csproj [hunk 9]
```diff
@@ -1,5 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.props" Condition="Exists('..\vowpalwabbit\packages\Antlr4.4.3.0\build\Antlr4.props')" />
```
Comment (JohnLangford) [inline cs/test/cs_test.csproj]: Delete?

File: cs/test/cs_test.csproj [hunk 10]
```diff
@@ -29,7 +30,12 @@
     <ApplicationVersion>1.0.0.%2a</ApplicationVersion>
     <UseApplicationTrust>false</UseApplicationTrust>
     <BootstrapperEnabled>true</BootstrapperEnabled>
+<<<<<<< HEAD:cs_test/cs_test.csproj
```
Comment (JohnLangford) [inline cs/test/cs_test.csproj]: Delete?

File: cs/test/cs_test.csproj [hunk 11]
```diff
@@ -29,7 +30,12 @@
...
+    <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\vowpalwabbit\</SolutionDir>
+    <NuGetPackageImportStamp>7bfb95b4</NuGetPackageImportStamp>
+=======
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\vowpalwabbit\</SolutionDir>
+>>>>>>> 684798fdd2eceb9abb108fe9e64a0c4ef5ef4a07:cs/test/cs_test.csproj
```
Comment (JohnLangford) [inline cs/test/cs_test.csproj]: Delete?

File: cs/test/cs_test.csproj [hunk 12]
```diff
@@ -153,6 +159,10 @@
     <CodeAnalysisIgnoreBuiltInRules>false</CodeAnalysisIgnoreBuiltInRules>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Antlr4.Runtime.net45, Version=4.3.0.0, Culture=neutral, PublicKeyToken=eb42632606e9261f, processorArchitecture=MSIL">
```
Comment (JohnLangford) [inline cs/test/cs_test.csproj]: Delete?

**Comment by JohnLangford** [comment]
There's a bunch of stuff in the diff which look like merge errors here: 

https://github.com/JohnLangford/vowpal_wabbit/pull/1007/files

Can you remove all the extra stuff unrelated to what we want to merge?

**Comment by haipengl** [comment]
Okay, I have removed all the extra stuff.

**Comment by JohnLangford** [comment]
1) Do we need RPNewton?
2) update_A() and check() allocate memory and then deallocate it.  If you preallocate the memory and then deallocate it at the end, things will work faster.   Can you do that?
3) There is no deallocation at the end.  You need a function like this: https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/bfgs.cc#L808 to avoid leaking memory.
4) After you do (3), can you test with valgrind to make sure there are no more memory leaks?

**Comment by haipengl** [comment]
Okay, I have removed RPNewton and added deallocation. Valgrind test also looked good.

**Comment by JohnLangford** [comment]
This is merged in now, thanks Haipeng.

All, we now have second order online learning working as a base learner as
per this paper: http://arxiv.org/abs/1602.02202 . Empirically, we found
this exceptional on only a few difficult datasets, but it's a nice hammer
to have available.

Haipeng, it may be worthwhile to separate out the sparse online pca at some
point---I expect people will find that generally useful.

-John

**Comment by haipengl** [comment]
Yes, separating out the sparse Oja part is in my todo list...

**Pull request merged**